pkg/agent/packetcapture/capture/bpf.go

```diff
@@ -135,28 +136,64 @@ func calculateSkipFalse(transport *transportFilters) uint8 {
 }
 
 // Generates IP address and port matching instructions
-func compileIPAndTransportFilters(srcAddrVal, dstAddrVal uint32, size, curLen uint8, transport *transportFilters, needsOtherTrafficDirectionCheck bool) []bpf.Instruction {
+func compileIPFilters(srcIP, dstIP net.IP, size, curLen uint8, transport *transportFilters, needsOtherTrafficDirectionCheck bool) []bpf.Instruction {
 	inst := []bpf.Instruction{}
 
 	// from here we need to check the inst length to calculate skipFalse. If no protocol is set, there will be no related bpf instructions.
 
+	// Calculate uint32 values from net.IP when needed
+	var srcAddrVal, dstAddrVal uint32
+	if srcIP != nil {
+		srcAddrVal = binary.BigEndian.Uint32(srcIP[len(srcIP)-4:])
+	}
+	if dstIP != nil {
+		dstAddrVal = binary.BigEndian.Uint32(dstIP[len(dstIP)-4:])
+	}
+
 	// calculate skip size to jump to the final instruction (NO MATCH)
 	skipToEnd := func() uint8 {
 		return size - curLen - uint8(len(inst)) - 2
 	}
 
-	// needsOtherTrafficDirectionCheck indicates if we need to check whether the packet belongs to the return traffic flow when source IP from the
-	// packet spec and packet header don't match and we are capturing packets in both direction. If true, we calculate skipFalse to jump to the
-	// instruction that compares the destination IP from the packet spec with the loaded source IP from the packet header.
-	if needsOtherTrafficDirectionCheck {
-		inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: 0, SkipFalse: calculateSkipFalse(transport)})
-	} else {
-		inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: 0, SkipFalse: skipToEnd()})
+	if srcIP != nil {
+		inst = append(inst, loadIPv4SourceAddress)
+		// needsOtherTrafficDirectionCheck indicates if we need to check whether the packet belongs to the
+		// return traffic flow when source IP from the packet spec and packet header don't match and we are
+		// capturing packets in both direction. If true, we calculate skipFalse to jump to the instruction
+		// that compares the destination IP from the packet spec with the loaded source IP from the packet
+		// header.
+		if needsOtherTrafficDirectionCheck {
+			inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: 0, SkipFalse: calculateSkipFalse(srcIP, dstIP, transport)})
+		} else {
+			inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: 0, SkipFalse: skipToEnd()})
+		}
+	}
+
+	if dstIP != nil {
+		inst = append(inst, loadIPv4DestinationAddress)
+		// When capturing in 'Both' directions and only a dstIP is specified (srcIP == nil),
```

Comment:
The comment talks about the `Both` direction, but there is no dependency on the direction in the code below, so that's confusing.
I feel like this comment is missing the real point of the `srcIP` check (or maybe it's just me misunderstanding): if `srcIP` is specified and we get to that point (`dstIP` check), then that means the source IP was a match; if the source IP was a match and the destination IP is not a match, then we can skip to the end directly as there is no chance of a match in either direction.